File: dotnet/Lexy.Poc/Lexy.Poc.Core/Parser/ValidationContext.cs
using System;
using System.Collections.Generic;
using Lexy.Poc.Core.Language;

namespace Lexy.Poc.Core.Parser
{
    public class ValidationContext : IValidationContext
    {
        private class CodeContextScope : IDisposable
        {
            private readonly Func<IFunctionCodeContext> func;

            public CodeContextScope(Func<IFunctionCodeContext> func) => this.func = func;

            public void Dispose() => func();
        }

        private readonly Stack<IFunctionCodeContext> contexts = new Stack<IFunctionCodeContext>();

        public IParserContext ParserContext { get; }
        public IFunctionCodeContext FunctionCodeContext { get; private set; }
        public IParserLogger Logger => ParserContext.Logger;
        public Nodes Nodes => ParserContext.Nodes;

        public ValidationContext(IParserContext context)
        {
            ParserContext = context ?? throw new ArgumentNullException(nameof(context));
        }

        public IDisposable CreateCodeContextScope()
        {
            if (FunctionCodeContext != null)
            {
                contexts.Push(FunctionCodeContext);
            }

            FunctionCodeContext = new FunctionCodeContext(Logger, FunctionCodeContext);

            return new CodeContextScope(() =>
            {
                return FunctionCodeContext = contexts.Count == 0 ? null : contexts.Pop();
            });
        }
    }
}

File: typescipt/src/parser/validationContext.ts
import {IVariableContext, VariableContext} from "./variableContext";
import {Stack} from "../infrastructure/lambdaComparer";
import {IParserLogger} from "./IParserLogger";
import {RootNodeList} from "../language/rootNodeList";

class CodeContextScope  {
  private readonly func: () => IVariableContext | null;

  constructor(func: (() => IVariableContext | null)) {
    this.func = func;
  }

  [Symbol.dispose] = () => {
    this.func();
  }
}

export interface IValidationContext {
  logger: IParserLogger;
  rootNodes: RootNodeList;

  variableContext: IVariableContext | null;
  createVariableScope() : { [Symbol.dispose] };
}

export class ValidationContext implements IValidationContext {
   private contexts: Stack<IVariableContext> = new Stack<IVariableContext>()
   private variableContextValue: IVariableContext | null;

   public logger: IParserLogger;
   public rootNodes: RootNodeList;

   constructor(logger: IParserLogger, rootNodes: RootNodeList) {
     this.logger = logger;
     this.rootNodes = rootNodes;
   }

   public get variableContext(): IVariableContext {
     if (this.variableContextValue == null) throw new Error(`FunctionCodeContext not set.`);
     return this.variableContextValue;
   }

   public createVariableScope(): { [Symbol.dispose] } {
     if (this.variableContextValue != null) {
       this.contexts.push(this.variableContextValue);
     }

     this.variableContextValue = new VariableContext(this.logger, this.variableContextValue);

     return new CodeContextScope(() => {
       return this.variableContextValue = this.contexts.size() == 0 ? null : this.contexts.pop();
     });
   }
}
